Bug fix and changes

ShadowRanger opened the pull request:
- Added automatic config updater
- Only use UUIDFetcher when a player isn't online
- Fixed null UUID error
- Use new, non-deprecated Material identification system
- Improved some chat output messages (added color)
- Small optimizations
- General refactoring
- Added new searchender command code (not yet implemented, wip)

I just want to run this pull request by you lishid first before it gets merged. I made it so new configs have a different layout and don't use full capitals in the settings. I made it like this because generally YAML config file don't usually have capital letters in the keys so I thought it would be a good idea to make this change to keep with common YAML format standards. But I wanted to make sure you were happy with it before I went ahead and did anything. Let me know if you prefer the old way with capitals and I will happily change it. :)

lishid commented on src/main/java/com/lishid/openinv/ConfigUpdater.java:
```diff
@@ -0,0 +1,161 @@
+    public void checkForUpdates() {
+        if (isConfigOutdated()) {
+            plugin.getLogger().info("[Config] Update found! Performing update...");
+            updateConfig();
```
I'd suggest comparing against specific versions and running specific upgrade commands for each version.
Basically upgradeConfig1to2(), and in the future, upgradeConfig2to3(), etc. This ensures that you're not risking running upgrade 1 to 2 more than once and error out because the migration already happened.

Something like if(version < 2) { upgrade1to2(); }
Make sure you set version to 2 and not CONFIG_VERSION in the process.

ShadowRanger replied:
Yeah I agree, that would be a better way to do it. Done.

lishid commented on src/main/java/com/lishid/openinv/ConfigUpdater.java:
```diff
@@ -0,0 +1,161 @@
+
+    }
+
+    private Map<UUID, Boolean> updateAnyChestToggles() {
```
These three methods looks very very similar. We can probably combine these into a single one that takes a parameter of a ConfigurationSection or a String for the key to get from the config.

ShadowRanger replied:
Great idea, just made it use 1 method for getting the toggle statuses.

lishid commented on src/main/java/com/lishid/openinv/OpenInv.java:
```diff
@@ -141,27 +120,83 @@ public static Object getFromConfig(String data, Object defaultValue) {
+        }
+
+        String itemName = mainPlugin.getConfig().getString("items.open-inv", "STICK");
+        return Material.getMaterial(itemName);
```
Probably want a null pointer check here?

ShadowRanger replied:
I believe config.isSet already checks for null values.

```java
    public boolean isSet(String path) {
        Configuration root = this.getRoot();
        return root == null?false:(root.options().copyDefaults()?this.contains(path):this.get(path, (Object)null) != null);
    }
```

lishid replied:
I meant Material.getMaterial might return a null Material.

ShadowRanger replied:
Oh, oops. I'll add a check in now.

lishid commented on src/main/java/com/lishid/openinv/ConfigUpdater.java:
```diff
@@ -0,0 +1,161 @@
+        }
+
+        // Set the new config options
+        plugin.getConfig().set("config-version", CONFIG_VERSION);
```
Could probably set Config config = plugin.getConfig() and use config later on. Minor detail though.

ShadowRanger replied:
Alright, will do.

lishid commented:
I'm fine with YAML/config keys either way. Might be worth double checking the bukkitdev page for the instructions if these gets changed.

I'd also say it would be useful if you can run the migration against a previous/old config file and see how well the new file come out, and if there are any inconsistencies.

lishid commented:
Other than that, I'm really happy this code is being cleaned up. Now I can feel less ashamed of the code here. :)

ShadowRanger commented:
Alright, I made those changes. https://github.com/ShadowRanger/OpenInv/commit/55deabe56ba8730cb3ab560b4a49ed701dacb990
Let me know if you see anything I missed out.

ShadowRanger commented:
Okay I'm just about to test the config updater to make sure it works on old config. Will let you know when done. Shouldn't take long.

lishid commented on src/main/java/com/lishid/openinv/ConfigUpdater.java:
```diff
@@ -0,0 +1,135 @@
+        switch (getConfigVersion()) {
+            case 1:
+                updateConfig1To2();
+                break;
```
I'm imagining this will eventually become something like this, is that what you're thinking as well?

case 1:
 upgradeConfig1to2();
 //No break, falls through directly
case 2:
 upgradeConfig2to3();
etc...

ShadowRanger replied:
Yep, exactly. I thought using a switch would be a neat way to do it.

ShadowRanger commented:
Just tested the config updater and it looks good to me.

Old config.yml: http://pastebin.com/Px5fRx3x
New config.yml: http://pastebin.com/uG7hXmdE

lishid commented:
Hmm I'd be more curious to test the toggles. That's the more error-prone part I'd say.

ShadowRanger commented:
With the openinv item I can definitely add some output if the item name provided isn't valid. I definitely agree with that.

As for toggles, what part are you not sure about?

lishid commented:
Oh just generally testing that an old config with toggles saved for anychest/silentchest do get migrated correctly. If this goes out to the world and the migration screwed up, the old config toggles are lost forever (unless plugin config backups were a thing)

ShadowRanger commented:
Yeah I understand what you mean. Whilst on that note, 1 issue that I realized would be a problem for some people is using this plugin with their server in offline mode with the config update/migration. Due to the fact that currently the toggle statuses are saved with lowercased player names, the UUIDFetcher is unable to get the UUID of the player if their username contains uppercased letters because of the fetcher being case sensetive.

As for config backups, that could definitely be possible. We could simply create a backup file of the old config.yml and name it "config_old.yml" so that if anything goes wrong people still have a backup to go back to if need be.

lishid commented:
That does sound problematic... I'm thinking of removing the catch in UUIDUtil and make it bubble up the exception. Then you can deal with the failure by not actually saving the config back and disable the plugin with an error message "Your server must be online to migrate the plugin from an old version".

Do you happen to know what bukkit currently does when you ask to get a player by name? Any chance we can use that as a fallback if querying UUIDs is not available?

ShadowRanger commented:
Yeah I can do something like that with the disabling of the plugin.

As for how Bukkit handles player retrieval by name:
https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/src/main/java/org/bukkit/craftbukkit/CraftServer.java#1243

```java
    @Override
    @Deprecated
    public OfflinePlayer getOfflinePlayer(String name) {
        Validate.notNull(name, "Name cannot be null");

        // If the name given cannot ever be a valid username give a dummy return, for scoreboard plugins
        if (!validUserPattern.matcher(name).matches()) {
            return new CraftOfflinePlayer(this, new GameProfile(invalidUserUUID, name));
        }

        OfflinePlayer result = getPlayerExact(name);
        if (result == null) {
            // This is potentially blocking :(
            GameProfile profile = MinecraftServer.getServer().getUserCache().getProfile(name);
            if (profile == null) {
                // Make an OfflinePlayer using an offline mode UUID since the name has no profile
                result = getOfflinePlayer(new GameProfile(UUID.nameUUIDFromBytes(("OfflinePlayer:" + name).getBytes(Charsets.UTF_8)), name));
            } else {
                // Use the GameProfile even when we get a UUID so we ensure we still have a name
                result = getOfflinePlayer(profile);
            }
        } else {
            offlinePlayers.remove(result.getUniqueId());
        }

        return result;
    }
```

I guess we could use that as a fallback. Just might not be as accurate but if it's the only option then might be worth it.

lishid commented:
Seems like an OK fallback, it looks through both online and offline users. In the case it doesn't find one I think we're ok throwing it out suppose that we can't connect to MC to find the UUID.

ShadowRanger commented:
Yeah, I think so. I'll do it tonight (within the next few hours) as I have to do some things in real life first. :)

ShadowRanger commented:
Alright I just pushed another commit. The ConfigUpdater now creates a backup of the old config.yml just in case something goes wrong. Additionally, the UUIDFetcher is now case insensetive. It can now retrieve player UUID's by their name without requiring exact upper/lower casing.

lishid commented:
Looking good!

ShadowRanger commented:
Are you happy with it being merged yet? Or is there anything else you'd like me to add or change? I'm just eager to release it because some people are complaining of it not working and I fixed the problem which was with the UUID fetching.

lishid commented:
Oh totally! If you're confident the migration is ready, I'm fine with next steps.

ShadowRanger commented:
Alright cool well what I plan on doing is running a few more tests today and just making sure everything is working okay as well as some finishing touches and then i'll release it on BukkitDev :)

lishid commented:
Cool cool!

ShadowRanger commented:
Alright just added another commit to this pull request. Did a bit more refactoring as well as added a new searchender command and added more outputting for config conversion.

I'm pretty sure that this pull request is ready to be merged however I'd like it if you could please just go over it to make sure there are no mistakes or I have missed out anything.

With the config migration part, I just want to make sure that you're happy with how it performs the update. As in, backing up the old config and then converting all possible usernames to UUID's and so on so forth.

Other than that I think it's almost good to go.

lishid commented:
Looks good to me.

ShadowRanger commented:
So someone is asking on the BukkitDev page whether we can have a version
compatible with 1.7.10. He says he is willing to sacrifice $10 and I don't
know what to tell him.

On Thu, Jun 25, 2015 at 4:38 AM, Lishid <notifications@github.com> wrote:

> Looks good to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lishid/OpenInv/pull/27#issuecomment-114975550>.
>

ShadowRanger commented:
I could create a separate GitHub branch maybe for 1.7.10, not sure.

ShadowRanger commented:
Disregard the last comments I made on here that I deleted. I changed my mind about it and don't think we need to make a whole separate version for 1.7.10 compatibility.

lishid commented:
We could release a single version for earlier version compatibility if the work involves minimal changes to the code other than swapping out the craftbukkit jar. Not totally sure that's the case, but either way I wouldn't want to promise maintaining a compatible version in the future.

lishid commented:
Wait, 1.7.10 is super old already, right?  It's like a year old or something...

ShadowRanger commented:
Yeah it is very old that's why I changed my mind and now don't think we should do that.

lishid commented:
That's fair enough.

ShadowRanger commented:
Plus the guy who asked for it just commented calling the people on the comments 'assholes' in Russian.
